Importing the necessary libraries and reading in the data

In [283]:
# importing the necessary libraries
import numpy as np
import pandas as pd

In [263]:
# reading in the data
df= pd.read_csv("./Dataset-radassist_synthetic_backlog_messy_headers.csv", skipinitialspace = True)
df.head()

Study ID PatientId   Modality  Body-Part PRIORITY requestDate   Scan Date  \
0   S00000    P93810          CT     Chest   Urgent  2025-08-01  2025-08-08   
1   S00001    P21395  Ultrasound     Chest  Routine  2025-07-28  2025-08-04   
2   S00002    P64987         MRI     Spine  Routine  2025-10-12  2025-11-08   
3   S00003    P54597       X-ray   Abdomen  Routine  2025-10-06  2025-10-16   
4   S00004    P70217          CT     Spine  Routine  2025-08-07  2025-09-02   

   ReportDate followUpRecommended follow up type followUpDueDate  \
0  2025-08-10               False           None             tbc   
1  2025-08-04               False           None             NaN   
2         NaN               False           None             NaN   
3  2025-10-17                True             CT      2025-11-28   
4  2025-09-07                   N           None             NaN   

  FollowUpStatus referringSpecialty     site Radiologist  \
0   Not required            Surgery  Trust C    Dr Brown   
1   Not required            Surgery  Trust C    Dr Brown   
2   Not required           Oncology  Trust C    Dr Patel   
3      Completed                 ED  Trust B    Dr Smith   
4   Not required            Surgery  Trust A    Dr Smith   

                           reportTextExcerpt  
0          CT Chest – no follow-up required.  
1  Ultrasound Chest – no follow-up required.  
2         MRI Spine – no follow-up required.  
3     X-ray Abdomen – follow-up recommended.  
4          CT Spine – no follow-up required.

In [264]:
df.shape

(510, 16)

Cleaning and fixing the column headers

In [265]:
df.columns

Index(['Study ID', 'PatientId', 'Modality ', 'Body-Part', 'PRIORITY',
       'requestDate', 'Scan Date', 'ReportDate', 'followUpRecommended',
       'follow up type', 'followUpDueDate', 'FollowUpStatus',
       'referringSpecialty', 'site', 'Radiologist', 'reportTextExcerpt'],
      dtype='object')

In [266]:
# fixing the headers
df.columns = (
    df.columns.str.strip()
              .str.replace(' ', '_')
              .str.replace('-', '_')
              .str.replace('/', '_')
              .str.replace('__', '_')
              .str.lower()
)
df.head()

study_id patientid    modality body_part priority requestdate   scan_date  \
0   S00000    P93810          CT     Chest   Urgent  2025-08-01  2025-08-08   
1   S00001    P21395  Ultrasound     Chest  Routine  2025-07-28  2025-08-04   
2   S00002    P64987         MRI     Spine  Routine  2025-10-12  2025-11-08   
3   S00003    P54597       X-ray   Abdomen  Routine  2025-10-06  2025-10-16   
4   S00004    P70217          CT     Spine  Routine  2025-08-07  2025-09-02   

   reportdate followuprecommended follow_up_type followupduedate  \
0  2025-08-10               False           None             tbc   
1  2025-08-04               False           None             NaN   
2         NaN               False           None             NaN   
3  2025-10-17                True             CT      2025-11-28   
4  2025-09-07                   N           None             NaN   

  followupstatus referringspecialty     site radiologist  \
0   Not required            Surgery  Trust C    Dr Brown   
1   Not required            Surgery  Trust C    Dr Brown   
2   Not required           Oncology  Trust C    Dr Patel   
3      Completed                 ED  Trust B    Dr Smith   
4   Not required            Surgery  Trust A    Dr Smith   

                           reporttextexcerpt  
0          CT Chest – no follow-up required.  
1  Ultrasound Chest – no follow-up required.  
2         MRI Spine – no follow-up required.  
3     X-ray Abdomen – follow-up recommended.  
4          CT Spine – no follow-up required.

Checking and renaming the columns

In [267]:
df.columns

Index(['study_id', 'patientid', 'modality', 'body_part', 'priority',
       'requestdate', 'scan_date', 'reportdate', 'followuprecommended',
       'follow_up_type', 'followupduedate', 'followupstatus',
       'referringspecialty', 'site', 'radiologist', 'reporttextexcerpt'],
      dtype='object')

In [268]:
# renaming the columns for consistency
df = df.rename(columns={
    "patientid": "patient_id",
    "requestdate": "request_date",
    "reportdate": "report_date",
    "followuprecommended": "follow_up_recommended",
    "followupduedate": "follow_up_due_date",
    "followupstatus": "follow_up_status",
    "referringspecialty": "referring_specialty",
    "reporttextexcerpt": "report_text_excerpt"
})
df.columns

Index(['study_id', 'patient_id', 'modality', 'body_part', 'priority',
       'request_date', 'scan_date', 'report_date', 'follow_up_recommended',
       'follow_up_type', 'follow_up_due_date', 'follow_up_status',
       'referring_specialty', 'site', 'radiologist', 'report_text_excerpt'],
      dtype='object')

In [269]:
df.head()

study_id patient_id    modality body_part priority request_date   scan_date  \
0   S00000     P93810          CT     Chest   Urgent   2025-08-01  2025-08-08   
1   S00001     P21395  Ultrasound     Chest  Routine   2025-07-28  2025-08-04   
2   S00002     P64987         MRI     Spine  Routine   2025-10-12  2025-11-08   
3   S00003     P54597       X-ray   Abdomen  Routine   2025-10-06  2025-10-16   
4   S00004     P70217          CT     Spine  Routine   2025-08-07  2025-09-02   

  report_date follow_up_recommended follow_up_type follow_up_due_date  \
0  2025-08-10                 False           None                tbc   
1  2025-08-04                 False           None                NaN   
2         NaN                 False           None                NaN   
3  2025-10-17                  True             CT         2025-11-28   
4  2025-09-07                     N           None                NaN   

  follow_up_status referring_specialty     site radiologist  \
0     Not required             Surgery  Trust C    Dr Brown   
1     Not required             Surgery  Trust C    Dr Brown   
2     Not required            Oncology  Trust C    Dr Patel   
3        Completed                  ED  Trust B    Dr Smith   
4     Not required             Surgery  Trust A    Dr Smith   

                         report_text_excerpt  
0          CT Chest – no follow-up required.  
1  Ultrasound Chest – no follow-up required.  
2         MRI Spine – no follow-up required.  
3     X-ray Abdomen – follow-up recommended.  
4          CT Spine – no follow-up required.

Checking the values in the data for inconsistent/missing entries and standardizing the values and replacing all inconsistent entries.

In [270]:
# checking for incosistent entries
cols = ["modality", "body_part", "priority", "follow_up_recommended", "follow_up_status", "referring_specialty", "site", "radiologist"]

for col in cols:
    print(f"\nUnique values in '{col}':")
    print(df[col].unique())


Unique values in 'modality':
['CT' 'Ultrasound' 'MRI' 'X-ray']

Unique values in 'body_part':
['Chest' 'Spine' 'Abdomen' 'Shoulder' 'Pelvis' 'Head' 'Knee' nan]

Unique values in 'priority':
['Urgent' 'Routine' 'ROUTINE' '2WW' nan 'Emergency' 'URGENT' 'EMERGENCY']

Unique values in 'follow_up_recommended':
['False' 'True' 'N' 'Yes' 'Y' 'TRUE' 'No' 'FALSE']

Unique values in 'follow_up_status':
['Not required' 'Completed' 'overdue' 'not required' 'Pending' 'scheduled'
 'Overdue' 'Scheduled' 'pending' 'completed']

Unique values in 'referring_specialty':
['Surgery' 'Oncology' 'ED' 'Orthopaedics' 'GP' 'Respiratory']

Unique values in 'site':
['Trust C' 'Trust B' 'Trust A']

Unique values in 'radiologist':
['Dr Brown' 'Dr Patel' 'Dr Smith' 'Dr Jones' 'Dr Lee']


In [271]:
# replacing the inconsistent values in the "priority" column
df["priority"] = df["priority"].replace({
    "ROUTINE": "Routine",
    "URGENT": "Urgent",
    "2WW": "Two-week wait"
})
df["priority"].unique()

array(['Urgent', 'Routine', 'Two-week wait', nan, 'Emergency',
       'EMERGENCY'], dtype=object)

In [272]:
# replacing the inconsistent values in the "follow_up_recommended" column
df["follow_up_recommended"] = df["follow_up_recommended"].replace({
    "Y": "True",
    "Yes": "True",
    "TRUE": "True",
    "FALSE": "False",
    "No": "False",
    "N": "False"
})
df["follow_up_recommended"].unique()

array(['False', 'True'], dtype=object)

In [273]:
# replacing the incosistent values in the "follow_up_status" column
df["follow_up_status"] = df["follow_up_status"].replace({
    "not required": "Not required",
    "overdue": "Overdue",
    "pending": "Pending",
    "completed": "Completed",
    "scheduled": "Scheduled"
})
df["follow_up_status"].unique()

array(['Not required', 'Completed', 'Overdue', 'Pending', 'Scheduled'],
      dtype=object)

In [274]:
df.head()

study_id patient_id    modality body_part priority request_date   scan_date  \
0   S00000     P93810          CT     Chest   Urgent   2025-08-01  2025-08-08   
1   S00001     P21395  Ultrasound     Chest  Routine   2025-07-28  2025-08-04   
2   S00002     P64987         MRI     Spine  Routine   2025-10-12  2025-11-08   
3   S00003     P54597       X-ray   Abdomen  Routine   2025-10-06  2025-10-16   
4   S00004     P70217          CT     Spine  Routine   2025-08-07  2025-09-02   

  report_date follow_up_recommended follow_up_type follow_up_due_date  \
0  2025-08-10                 False           None                tbc   
1  2025-08-04                 False           None                NaN   
2         NaN                 False           None                NaN   
3  2025-10-17                  True             CT         2025-11-28   
4  2025-09-07                 False           None                NaN   

  follow_up_status referring_specialty     site radiologist  \
0     Not required             Surgery  Trust C    Dr Brown   
1     Not required             Surgery  Trust C    Dr Brown   
2     Not required            Oncology  Trust C    Dr Patel   
3        Completed                  ED  Trust B    Dr Smith   
4     Not required             Surgery  Trust A    Dr Smith   

                         report_text_excerpt  
0          CT Chest – no follow-up required.  
1  Ultrasound Chest – no follow-up required.  
2         MRI Spine – no follow-up required.  
3     X-ray Abdomen – follow-up recommended.  
4          CT Spine – no follow-up required.

Checking and changing the columns with the wrong data types with the correct data types

In [275]:
# checking the data types
df.dtypes

study_id                 object
patient_id               object
modality                 object
body_part                object
priority                 object
request_date             object
scan_date                object
report_date              object
follow_up_recommended    object
follow_up_type           object
follow_up_due_date       object
follow_up_status         object
referring_specialty      object
site                     object
radiologist              object
report_text_excerpt      object
dtype: object

In [276]:
# replacing the placeholders (eg: :"tbc") with actual with proper missing values for data consistency
df['follow_up_due_date'] = df['follow_up_due_date'].replace(
    ['tbc', 'TBC', 'n/a', 'N/A', 'none', 'None', '?', ''], np.nan
)

# extracting only the dates
df['request_date'] = df['request_date'].astype(str).str.split(' ').str[0]
df['scan_date'] = df['scan_date'].astype(str).str.split(' ').str[0]
df['report_date'] = df['report_date'].astype(str).str.split(' ').str[0]
df['follow_up_due_date'] = df['follow_up_due_date'].astype(str).str.split(' ').str[0]

# converting all the data types of all the date columns to datetime
df['request_date'] = pd.to_datetime(df['request_date'], errors='coerce')
df['scan_date'] = pd.to_datetime(df['scan_date'], errors='coerce')
df['report_date'] = pd.to_datetime(df['report_date'], errors='coerce')
df['follow_up_due_date'] = pd.to_datetime(df['follow_up_due_date'], errors='coerce')

In [277]:
df.dtypes

study_id                         object
patient_id                       object
modality                         object
body_part                        object
priority                         object
request_date             datetime64[ns]
scan_date                datetime64[ns]
report_date              datetime64[ns]
follow_up_recommended            object
follow_up_type                   object
follow_up_due_date       datetime64[ns]
follow_up_status                 object
referring_specialty              object
site                             object
radiologist                      object
report_text_excerpt              object
dtype: object

In [278]:
df.head()

study_id patient_id    modality body_part priority request_date  scan_date  \
0   S00000     P93810          CT     Chest   Urgent   2025-08-01 2025-08-08   
1   S00001     P21395  Ultrasound     Chest  Routine   2025-07-28 2025-08-04   
2   S00002     P64987         MRI     Spine  Routine   2025-10-12 2025-11-08   
3   S00003     P54597       X-ray   Abdomen  Routine   2025-10-06 2025-10-16   
4   S00004     P70217          CT     Spine  Routine   2025-08-07 2025-09-02   

  report_date follow_up_recommended follow_up_type follow_up_due_date  \
0  2025-08-10                 False           None                NaT   
1  2025-08-04                 False           None                NaT   
2         NaT                 False           None                NaT   
3  2025-10-17                  True             CT         2025-11-28   
4  2025-09-07                 False           None                NaT   

  follow_up_status referring_specialty     site radiologist  \
0     Not required             Surgery  Trust C    Dr Brown   
1     Not required             Surgery  Trust C    Dr Brown   
2     Not required            Oncology  Trust C    Dr Patel   
3        Completed                  ED  Trust B    Dr Smith   
4     Not required             Surgery  Trust A    Dr Smith   

                         report_text_excerpt  
0          CT Chest – no follow-up required.  
1  Ultrasound Chest – no follow-up required.  
2         MRI Spine – no follow-up required.  
3     X-ray Abdomen – follow-up recommended.  
4          CT Spine – no follow-up required.

Checking and dropping the duplicate data

In [279]:
# checking how many duplicate values we have in the data
df.duplicated().sum()

10

In [ ]:
# dropping the duplicated values
df = df.drop_duplicates(subset=['study_id','patient_id'], keep='first')
df.shape 

(500, 16)

In [281]:
df.head()

study_id patient_id    modality body_part priority request_date  scan_date  \
0   S00000     P93810          CT     Chest   Urgent   2025-08-01 2025-08-08   
1   S00001     P21395  Ultrasound     Chest  Routine   2025-07-28 2025-08-04   
2   S00002     P64987         MRI     Spine  Routine   2025-10-12 2025-11-08   
3   S00003     P54597       X-ray   Abdomen  Routine   2025-10-06 2025-10-16   
4   S00004     P70217          CT     Spine  Routine   2025-08-07 2025-09-02   

  report_date follow_up_recommended follow_up_type follow_up_due_date  \
0  2025-08-10                 False           None                NaT   
1  2025-08-04                 False           None                NaT   
2         NaT                 False           None                NaT   
3  2025-10-17                  True             CT         2025-11-28   
4  2025-09-07                 False           None                NaT   

  follow_up_status referring_specialty     site radiologist  \
0     Not required             Surgery  Trust C    Dr Brown   
1     Not required             Surgery  Trust C    Dr Brown   
2     Not required            Oncology  Trust C    Dr Patel   
3        Completed                  ED  Trust B    Dr Smith   
4     Not required             Surgery  Trust A    Dr Smith   

                         report_text_excerpt  
0          CT Chest – no follow-up required.  
1  Ultrasound Chest – no follow-up required.  
2         MRI Spine – no follow-up required.  
3     X-ray Abdomen – follow-up recommended.  
4          CT Spine – no follow-up required.

Finally saving the data in a .CSV file

In [282]:
# saving the data in a csv file
df.to_csv('radassist_backlog_clean.csv', index=False)

Data Quality Summary:
- Rows before cleaning: 510
- Rows after cleaning: 500
- Duplicates removed: 10

Nulls handled (by column)
- Non-date placeholders (e.g., “tbc”) in the "follow_up_due_date" column were replaced with proper missing datetime values (NaT) for consistency.

- Several categorical columns such as "follow_up_type", "body_part", "priority" contained "None" and "Nan" values. These were not modified, as pandas automatically interprets None as missing in object-type columns. Leaving them unchanged keeps the dataset’s integrity while still allowing them to be recognized as missing values in any analysis.

Notes / assumptions:

- Only duplicates based on study_id + patient_id were removed.

- Missing categorical values "None" and "Nan" were left as valid missing indicators rather than dropped.

- Date placeholders like "tbc" and "N/A" were converted to NaT during datetime processing.

- All column names were standardized to snake_case.

- The values in the "follow_up_recommended" column was normalized to have boolean (True/False) values.